dnstable_convert.c

```diff
@@ -605,34 +658,35 @@ process_rdata_name_rev(Nmsg__Sie__DnsDedupe *dns, size_t i, ubuf *key, ubuf *val
 		return;		/* Other rrtypes are not indexed by name. */
 	}
 
-	if (dns->rdata[i].len == 0 || dns->rdata[i].len <= offset)
+	if (rdata_len <= offset)
 		return;
 
 	/* check for a len that is longer than the data left in the packet */
-	if (offset + len > dns->rdata[i].len)
+	if (offset + len > rdata_len)
 		return;
 
-	/* clear key, val */
-	ubuf_clip(key, 0);
+	/* value: the RRtype */
 	ubuf_clip(val, 0);
+	put_rrtype(dns, val);
 
-	/* key: type byte */
-	ubuf_add(key, ENTRY_TYPE_RDATA_NAME_REV);
+	rdata_name_rev_add(dns, key, val, rdata_start + offset, len, do_downcase);
 
-	/* key: rdata name (label-reversed) */
-	res = wdns_reverse_name(dns->rdata[i].data + offset, len, name);
-	assert(res == wdns_res_success);
-	if (do_downcase) {
-		downcase.data = name;
-		downcase.len = len;
-		wdns_downcase_name(&downcase);
-	}
-	ubuf_append(key, name, len);
+	/* For SOA RDATA, always process the MNAME, and then the RNAME if requested. */
+	if (dns->rrtype == WDNS_TYPE_SOA && s_process_soa_rname) {
+		offset = len;		/* Offset of RNAME (== length of MNAME). */
+		res = wdns_len_uname(rdata_start + offset, rdata_end, &len);
+		if (res != wdns_res_success)
+			return;
 
-	/* value: the RRtype */
-	put_rrtype(dns, val);
+		if (offset + len > rdata_len)
+			return;
 
-	add_entry(dns, key, val);
+		/* Do not duplicate if MNAME and RNAME are the same. */
+		if (offset == len && memcmp(rdata_start, rdata_start + offset, len) == 0)
```

Comment:
I'm not sure this check is really necessary but it ultimately doesn't matter much.

Reply:
This comparison needs to be made in `process_rdata_slice()` so that an SOA entry with mname == rname is not recorded in both sliced and unsliced form, as this will lead to its duplication in all matching searches.

Reply:
Done.